feat:Add loadbalancer

Yuukadesu opened the pull request:
<!--
  ~ Licensed to the Apache Software Foundation (ASF) under one or more
  ~ contributor license agreements.  See the NOTICE file distributed with
  ~ this work for additional information regarding copyright ownership.
  ~ The ASF licenses this file to You under the Apache License, Version 2.0
  ~ (the "License"); you may not use this file except in compliance with
  ~ the License.  You may obtain a copy of the License at
  ~
  ~    http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
  ~
  -->

  <!--
Thanks for contributing! Here are some tips you can follow to help us incorporate your contribution quickly and easily:
1. If this is your first time, please read our contributor guidelines:
    - https://streampipes.apache.org/community/get-involved/
    - https://cwiki.apache.org/confluence/display/STREAMPIPES/Getting+Started
2. Make sure the PR title is formatted like: `[#<GitHub issue id>] PR title ...`
3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., `[WIP][#<GitHub issue id>] PR title ...`.
4. Please write your PR title to summarize what this PR proposes/fixes.
5. Link the PR to the corresponding GitHub issue (if present) in the `Development` section in the right menu bar.
6. Be sure to keep the PR description updated to reflect all changes.
7. If possible, provide a concise example to reproduce the issue for a faster review.
8. Make sure tests pass via `mvn clean install`.
9. (Optional) If the contribution is large, please file an Apache ICLA
    - http://apache.org/licenses/icla.pdf
-->

### Purpose
<!--
Please clarify what changes you are proposing and describe how those changes will address the issue.
Furthermore, describe potential consequences the changes might have.
-->

### Remarks
<!--
Is there anything left we need to pay attention on?
Are there some references that might be important? E.g. links to Confluence, or discussions
on the mailing list or GitHub.
-->
PR introduces (a) breaking change(s): <yes/no>

PR introduces (a) deprecation(s): <yes/no>

Yuukadesu committed:
feat:Add loadbalancer

Yuukadesu committed:
update

Yuukadesu committed:
update

Yuukadesu committed:
update code

Yuukadesu committed:
update code

Yuukadesu committed:
Merge branch 'dev' into load-balance

Yuukadesu committed:
update code

Yuukadesu committed:
Merge remote-tracking branch 'origin/load-balance' into load-balance

luoluoyuyu commented on streampipes-commons/src/main/java/org/apache/streampipes/commons/constants/Envs.java:
```diff
@@ -84,13 +84,13 @@ public enum Envs {
   SP_LOG_FETCH_INTERVAL_MS("SP_LOG_FETCH_INTERVAL_MS", "60000"),
 
-  SP_HEALTH_SERVICE_MAX_UNHEALTHY_TIME_MS("SP_HEALTH_SERVICE_MAX_UNHEALTHY_TIME_MS", "60000"),
+  SP_HEALTH_SERVICE_MAX_UNHEALTHY_TIME_MS("SP_HEALTH_SERVICE_MAX_UNHEALTHY_TIME_MS", "1000"),
```
The frequency is too high, use the original value

luoluoyuyu commented on streampipes-extensions-api/src/main/java/org/apache/streampipes/extensions/api/locker/Lock.java:
```diff
@@ -0,0 +1,44 @@
+ * Interface for pipeline locking operations.
+ * Provides methods to acquire and release locks for pipeline processing.
+ */
+public interface Lock {
```
delete

luoluoyuyu commented on streampipes-model/src/main/java/org/apache/streampipes/model/extensions/svcdiscovery/SpServiceTagPrefix.java:
```diff
@@ -23,9 +23,16 @@ public enum SpServiceTagPrefix {
-  DATA_SINK;
+  DATA_SINK,
+  SYSTEM;
+
```
Why add this tag

luoluoyuyu commented on streampipes-model/src/main/java/org/apache/streampipes/model/loadbalancer/PipelineInfo.java:
```diff
@@ -0,0 +1,140 @@
+  private PipelineStates pipelineState;
+  private Thread currentThread;
+  private LocalDateTime lockTime;
+  private Integer waitCount;
```
delete

luoluoyuyu commented on streampipes-model/src/main/java/org/apache/streampipes/model/loadbalancer/PipelineStates.java:
```diff
@@ -0,0 +1,61 @@
+  /**
+   * Pipeline is stopped.
+   */
+  STOP,
```
delete

luoluoyuyu commented on streampipes-pipeline-management/src/main/java/org/apache/streampipes/manager/health/ServiceHealthCheck.java:
```diff
@@ -97,4 +113,19 @@ private String makeHealthCheckUrl(SpServiceRegistration service) {
+    return activeServices
+            .stream()
+            .filter(service -> filtersSupported(service, tag))
+            .filter(service -> service.getStatus() != SpServiceStatus.UNHEALTHY)
```
service -> service.getStatus() == SpServiceStatus.HEALTHY

luoluoyuyu commented on streampipes-pipeline-management/src/main/java/org/apache/streampipes/manager/health/ServiceRegistrationManager.java:
```diff
@@ -69,6 +69,11 @@ public List<SpServiceRegistration> getAllServices() {
     return storage.findAll();
   }
 
+  public List<SpServiceRegistration> getAivServices() {
```
getActiveService

luoluoyuyu commented on streampipes-pipeline-management/src/main/java/org/apache/streampipes/manager/loadbalance/LoadManager.java:
```diff
@@ -0,0 +1,198 @@
+    ExtensionServiceSelector selector;
+    PipelineMigrator migrator;
+
+    LoadBalancerConfig.LoadTargetStd = environment.getLoadTargetStd().getValueOrDefault();
```
New configuration needs to be added. Users may not want to enable dynamic load balancing, and the current method is unstable. Please set ```LoadManageEnable``` to false

luoluoyuyu commented on streampipes-pipeline-management/src/main/java/org/apache/streampipes/manager/loadbalance/unit/ResourceUnitScanner.java:
```diff
@@ -0,0 +1,347 @@
+    // Find pipelines using this service
+    List<Pipeline> relevantPipelines = allPipelines.stream()
+      .filter(pipeline -> pipelineUsesService(pipeline, serviceUrl))
+      .toList();
```
Please filter out Pipelines that are not in the running state

luoluoyuyu commented on streampipes-pipeline-management/src/main/java/org/apache/streampipes/manager/loadbalance/unit/ResourceUnitScanner.java:
```diff
@@ -0,0 +1,347 @@
+    // Find pipelines using this service
+    List<Pipeline> relevantPipelines = allPipelines.stream()
+      .filter(pipeline -> pipelineUsesService(pipeline, serviceUrl))
+      .toList();
```
Please filter out Pipelines that are not in the running state

luoluoyuyu commented on streampipes-pipeline-management/src/main/java/org/apache/streampipes/manager/loadbalance/unit/ResourceUnitScanner.java:
```diff
@@ -0,0 +1,347 @@
+    // Find adapters running on this service
+    List<AdapterDescription> serviceAdapters = allAdapters.stream()
+      .filter(adapter -> serviceUrl != null && serviceUrl.equals(adapter.getSelectedEndpointUrl()))
+      .toList();
```
Please filter out adapters that are not in the running state

luoluoyuyu commented on streampipes-pipeline-management/src/main/java/org/apache/streampipes/manager/loadbalance/unit/ResourceUnitScanner.java:
```diff
@@ -0,0 +1,347 @@
+    // Find adapters running on this service
+    List<AdapterDescription> serviceAdapters = allAdapters.stream()
+      .filter(adapter -> serviceUrl != null && serviceUrl.equals(adapter.getSelectedEndpointUrl()))
+      .toList();
```
Please filter out adapters that are not in the running state

Yuukadesu committed:
update code

luoluoyuyu commented on streampipes-commons/src/main/java/org/apache/streampipes/commons/environment/DefaultEnvironment.java:
```diff
@@ -15,15 +15,12 @@
-import org.apache.streampipes.commons.environment.variable.BooleanEnvironmentVariable;
-import org.apache.streampipes.commons.environment.variable.IntEnvironmentVariable;
-import org.apache.streampipes.commons.environment.variable.StringEnvironmentVariable;
+import org.apache.streampipes.commons.environment.variable.*;
```
import org.apache.streampipes.commons.environment.variable.BooleanEnvironmentVariable;
import org.apache.streampipes.commons.environment.variable.IntEnvironmentVariable;
import org.apache.streampipes.commons.environment.variable.StringEnvironmentVariable;

Yuukadesu committed:
update code

Yuukadesu committed:
Merge branch 'dev' into load-balance

Yuukadesu committed:
update code

Yuukadesu committed:
Merge remote-tracking branch 'origin/load-balance' into load-balance

Yuukadesu committed:
update code

Yuukadesu committed:
update code

Yuukadesu commented:
This PR has realized the discussion https://github.com/apache/streampipes/discussions/2552
In this PR, I added LoadBalancer to load balance StreamPipes.
Specifically:
First, three allocation algorithms are used: weighted random algorithm, minimum resource allocation rate algorithm and weight first allocation algorithm to achieve pipeline allocation.
After allocation, migration algorithms such as threshold migration, overload migration, and unified load migration will be used to perform pipeline migration operations when the program is running to ensure load balance of all extended services.
In addition, I added indicators to monitor the load balancer, which enables me to observe the number of adapters and pipelines for each service and the time spent on each migration operation.

Yuukadesu commented:
These are the displays of Pipeline_count, adapter_count, and migration_time that I added in Grafana.
Pipeline_count(There are two overlaps)
<img width="993" height="310" alt="截屏2025-10-28 21 22 11" src="https://github.com/user-attachments/assets/9daa7e93-444b-4ebf-93d2-596a35ba0b97" />
adapter_count
<img width="986" height="309" alt="截屏2025-10-28 21 21 15" src="https://github.com/user-attachments/assets/93a6b27b-ec31-4629-8c89-9371850053c4" />
migration_time
<img width="984" height="309" alt="截屏2025-10-28 21 22 46" src="https://github.com/user-attachments/assets/792a6a44-5170-407d-b211-50172575b07c" />

Yuukadesu commented:
The following is a test of the load balancing effect under the weight random allocation algorithm. I started 40 adapters and 40 pipelines for this test, and it can be seen that after a period of time, all indicators are basically the same.
CPU_Usage
<img width="987" height="304" alt="截屏2025-10-29 09 32 45" src="https://github.com/user-attachments/assets/20c49e35-a6f1-431d-8ae0-757044cdae2b" />
memory_usage
<img width="983" height="307" alt="截屏2025-10-29 09 33 00" src="https://github.com/user-attachments/assets/2e850fa4-ad23-483c-afc7-db7083387bbe" />
system_load
<img width="983" height="308" alt="截屏2025-10-29 09 33 22" src="https://github.com/user-attachments/assets/9fd7cf16-dc7b-4756-8338-75711151d18d" />
historical_system_load
<img width="982" height="304" alt="截屏2025-10-29 09 33 38" src="https://github.com/user-attachments/assets/a9c31e40-98aa-4adb-8f81-969dd9493c49" />
After running for a period of time, because of my computer cannot withstand this load, an extension service experienced a brief crash, and some pipelines also stopped, resulting in a decrease in load. I used ThresholdMigrator or to solve this problem.It can be seen that after a period of time, the system re balances the load.
CPU_Usage
<img width="987" height="302" alt="截屏2025-10-29 09 46 23" src="https://github.com/user-attachments/assets/05ee8dc7-9440-4af2-be11-16e20c6654a3" />
memory_usage
<img width="986" height="305" alt="截屏2025-10-29 09 48 03" src="https://github.com/user-attachments/assets/3b58cfcb-f614-4851-8c6e-c6191c3dcd94" />
system_load
<img width="987" height="308" alt="截屏2025-10-29 09 48 18" src="https://github.com/user-attachments/assets/cda04f89-ed8a-4a51-81ea-6e249cf2fbcc" />
historical_system_load
<img width="988" height="307" alt="截屏2025-10-29 09 48 45" src="https://github.com/user-attachments/assets/10f2047f-750c-4f31-b9d2-a1583c92ccf7" />

Yuukadesu commented:
The following is a test of the load balancing effect under the weight first allocation algorithm. I started 40 adapters and 40 pipelines for this test, and it can be seen that after a period of time, all indicators are basically the same.
CPU_Usage
![1631761669194_ pic](https://github.com/user-attachments/assets/28fd788d-ce10-4565-9575-8738d9091210)
Memory_Usage
![1641761669211_ pic](https://github.com/user-attachments/assets/82cbb91e-7113-437f-b7b3-544dab39ba5a)
system_load
![1651761669227_ pic](https://github.com/user-attachments/assets/b21282fd-a2c2-498a-b6d0-27ebe1eec65a)
historical_system_load
![1661761669240_ pic](https://github.com/user-attachments/assets/1606f619-67ef-4d16-ad4b-357d5fa34655)
Then I stopped one of the extension services to simulate a service outage during production and test the reallocation when a failure occurs.And I use TransferMigrator to do this.
CPU_Usage
![1671761669559_ pic](https://github.com/user-attachments/assets/05d5c2b2-fc96-4fba-91ad-bd626225d3a8)
Memory_Usage
![1681761669571_ pic](https://github.com/user-attachments/assets/8ac4c25a-39e9-4491-9f25-01b8489abb3e)
system_load
![1691761669596_ pic](https://github.com/user-attachments/assets/0a3c0cb2-5dc8-4ab8-bc2d-1382a2ee1207)
historical_system_load
![1701761669614_ pic](https://github.com/user-attachments/assets/b8eb1236-de43-4f4f-ac92-a72a56976068)

Yuukadesu commented:
I also added an option to enable load balancing function. When it is false, the program runs normally but does not perform load balancing.
CPU_Usage
<img width="990" height="313" alt="截屏2025-10-29 10 11 06" src="https://github.com/user-attachments/assets/0aca7a3a-ddcc-4e35-838c-8a5a34e1178e" />
memory_Usage
<img width="988" height="308" alt="截屏2025-10-29 10 11 30" src="https://github.com/user-attachments/assets/ee8c8999-51a1-4d28-a87c-3840e1892a7a" />
system_load
<img width="986" height="311" alt="截屏2025-10-29 10 11 48" src="https://github.com/user-attachments/assets/1f13eb3e-fd7a-4e95-bbde-504869a81f25" />
historical_system_load
<img width="986" height="303" alt="截屏2025-10-29 10 12 09" src="https://github.com/user-attachments/assets/b271a419-b010-4c63-a683-4ec08d5d9f05" />

luoluoyuyu commented:
Great, thanks @Yuukadesu  for completing the multi-extension service runtime adapter and pipeline extension.

dominikriemer commented:
@Yuukadesu @luoluoyuyu awesome, I'm really impressed what you've been working on here. Very cool feature!
Give me some days to try it and dig through the code - if there is any open issue where I can help please let me know.

luoluoyuyu commented:
Hi @dominikriemer , the load balancer has some configurations added as follows:
LOAD_MANAGER_ENABLE  is Master switch:  
- `true` (default) – ExtensibleLoadManager is instantiated at startup; all load-balancing and traffic-migration features are active.  
- `false` – Selector/Migrator initialization is skipped; dynamic load management is disabled and the service falls back to static or default routing.

| Configuration Key | Legal Value (case-sensitive) | Meaning |
|-------------------|------------------------------|---------|
| SELECTOR | WeightedRandomSelector | Weighted random |
| | MinimumLoadSelector | Least-loaded |
| | WeightedFirstSelector | Weighted first |
| MIGRATOR | TransferMigrator | Manual transfer |
| | OverloadMigrator | Overload-based migration |
| | ThresholdMigrator | Threshold-based migration |

luoluoyuyu commented:
@dominikriemer Thank you so much for the positive feedback on this feature! Given that the OSPP deadline is 31 October, could we kindly proceed with merging this PR if there are no critical concerns? We're fully committed to addressing any follow-ups immediately after the deadline.
Your support in accelerating this process would be immensely helpful!

dominikriemer commented:
Ok! I'll take a look tonight.

dominikriemer commented on streampipes-pipeline-management/src/main/java/org/apache/streampipes/manager/health/PipelineHealthCheck.java:
```diff
@@ -224,30 +222,25 @@ private void addEndpoint(Map<String, List<InvocableStreamPipesEntity>> endpointM
+    try {
+      this.checkAndRestorePipelineElements();
+    } catch (Exception e) {
+
```
Should we log the exception here?

dominikriemer commented on streampipes-rest-extensions/pom.xml:
```diff
@@ -75,6 +75,12 @@
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.apache.streampipes</groupId>
```
This module contains many things that are only used by the core service.
Can we extract the parts of the load balancer needed by both core and extension services into a separate module?
Then we could avoid importing too much functionality from the core into the extension services.

dominikriemer commented on streampipes-storage-api/src/main/java/org/apache/streampipes/storage/api/IPipelineStorage.java:
```diff
@@ -26,4 +26,13 @@ public interface IPipelineStorage extends CRUDStorage<Pipeline> {
 
+  void storePipeline(Pipeline pipeline);
+
+  void updatePipeline(Pipeline pipeline);
```
I think these methods can be removed and replaced by the methods offered by `CRUDStorage`.

dominikriemer reviewed (approved):
Great work @Yuukadesu @luoluoyuyu !
I had a look at the code and added just a few questions. We can discuss these also later after you've merged the code.
Really like this feature! :)

Yuukadesu committed:
update code

Yuukadesu committed:
update code

Yuukadesu committed:
Merge branch 'apache:dev' into load-balance

Yuukadesu committed:
update code

Yuukadesu committed:
update code
